### DISCUSSION TITLE
T3D spindle modbus control

### DISCUSSION THREAD

**Discussion by Nick507** [grblHAL/core#477] [Q&A]
Hi,
I have successfully connected T3D spindle, using modvfd plugin, but bumped with two issues:

1. get rpm command of this spindle returns signed 16 bit value, which is negative for cw and positive for ccw. My idea is to change f2rpm function argument to int16_t (from uint16_t) and add fabs in return. Is it good idea for pull request? This should not break other spindles with unsigned values, until 32767 rpm (I know that on milling machine spindles this is not a limit). Or the right way is to make new settings parameter?

2. stop command for T3D spindle - stops it without deceleration, so the right sequence in my case would be set zero speed, sync speed and then issue stop command. I thought, that this logic can be implemented in common part, like spindle_sync, or somewhere else. Or may be you can suggest another proper way to fix this.

I can write my own spindle driver, but I'm trying to avoid that.

Thank you in advance!

**Comment by terjeio** [comment]
Add [Setting_VFD_20](https://github.com/grblHAL/core/blob/d9ed1bd89cdd7ea5dff3a1b217912d2447edb0cc/settings.h#L302) with two flags, one for how to handle signed value and one for how stop with deceleration should then be handled?
A downside of adding settings is that driver settings will be reset to default on an update unless they can be added in "holes" in the setting struct, luckily it looks like 16 bits can be inserted in the [vfd_settings_t struct](https://github.com/grblHAL/Plugins_spindle/blob/9c949fb50521d7a99fd8e262b9a49a34f406731f/vfd/spindle.h#L74-L75) without causing it to change size...

**Reply by Nick507** [reply]
Settings flag for signed/unsigned speed value is good. 
I've doubts about deceleration flag on vfd settings level, because it requires waiting loop in spindleSetState, which as I understand was not intended to be blocking. Maybe add this as a spindle capability, and handle in spindle_sync ?
